client/lib/numalib/topology.go

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+// Package numalib provides information regarding the system NUMA, CPU, and
+// device topology of the system.
+//
+// https://docs.kernel.org/6.2/x86/topology.html
+package numalib
+
+import (
+	"fmt"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/nomad/client/lib/idset"
+)
+
+// CoreGrade describes whether a specific core is a performance or efficiency
+// core type. A performance core generally has a higher clockspeed and can do
+// more than an efficiency core.
+type CoreGrade bool
+
+const (
+	performance CoreGrade = true
+	efficiency  CoreGrade = false
+)
+
+func gradeOf(siblings *idset.Set[CoreID]) CoreGrade {
+	switch siblings.Size() {
+	case 0, 1:
+		return efficiency
+	default:
+		return performance
+	}
+}
+
+func (g CoreGrade) String() string {
+	switch g {
+	case performance:
+		return "performance"
+	default:
+		return "efficiency"
+	}
+}
+
+type (
+	NodeID   uint8
+	SocketID uint8
+	CoreID   uint16
+	KHz      uint64
+	MHz      uint64
+	GHz      float64
+	Cost     uint8
+)
+
+func (hz KHz) MHz() MHz {
+	return MHz(hz / 1000)
+}
+
+func (hz KHz) String() string {
+	return strconv.FormatUint(uint64(hz.MHz()), 10)
+}
```

Comment:
Calling this pointer receiver `hz` feels a little gross :grinning: 

```suggestion
func (khz KHz) MHz() MHz {
	return MHz(khz / 1000)
}

func (khz KHz) String() string {
	return strconv.FormatUint(uint64(khz.MHz()), 10)
}
```

Reply:
whoops! :scream: